javascript/packages/orchestrator/src/providers/native/resources/node.resource.ts

```diff
@@ -0,0 +1,145 @@
+  protected readonly dataPath: string;
+  private readonly relayDataPath: string;
+
+  constructor(
+    client: Client,
+    protected readonly namespace: string,
+    protected readonly nodeSetupConfig: Node,
+  ) {
+    const nodeName = nodeSetupConfig.name;
+    this.configPath = `${client.tmpDir}/${nodeName}/cfg`;
+    this.dataPath = `${client.tmpDir}/${nodeName}/data`;
+    this.relayDataPath = `${client.tmpDir}/${nodeName}/relay-data`;
+  }
+
+  public async generateSpec() {
+    await this.createDirectories();
+    const ports = await this.generatePorts();
+    const command = await this.generateCommand();
+    const zombieRole = this.getZombieRole();
+    const env = this.getEnv();
+    const nodeManifest = this.generateNodeSpec(ports, command, zombieRole, env);
+
+    return nodeManifest;
+  }
+
+  protected async createDirectories() {
+    await makeDir(this.configPath, true);
+    await makeDir(this.dataPath, true);
+    await makeDir(this.relayDataPath, true);
+  }
+
+  private async portFromNodeSetupConfigOrDefault(portProperty: PortProperty) {
+    if (this.nodeSetupConfig[portProperty]) {
+      return this.nodeSetupConfig[portProperty];
+    }
+
+    return getRandomPort();
+  }
+
+  private async generatePorts() {
+    return [
+      {
+        containerPort: PROMETHEUS_PORT,
+        name: "prometheus",
+        flag: "--prometheus-port",
+        hostPort: await this.portFromNodeSetupConfigOrDefault("prometheusPort"),
+      },
+      {
+        containerPort: RPC_HTTP_PORT,
+        name: "rpc",
+        flag: "--rpc-port",
+        hostPort: await this.portFromNodeSetupConfigOrDefault("rpcPort"),
+      },
+      {
+        containerPort: RPC_WS_PORT,
+        name: "rpc-ws",
+        flag: "--ws-port",
+        hostPort: await this.portFromNodeSetupConfigOrDefault("wsPort"),
+      },
+      {
+        containerPort: P2P_PORT,
+        name: "p2p",
+        flag: "--port",
+        hostPort: await this.portFromNodeSetupConfigOrDefault("p2pPort"),
+      },
+    ];
+  }
+
+  protected generateCommand() {
+    if (this.nodeSetupConfig.zombieRole === "cumulus-collator") {
+      return genCumulusCollatorCmd(
+        this.nodeSetupConfig,
+        this.configPath,
+        this.dataPath,
+        this.relayDataPath,
+        false,
+      );
+    }
+
+    return genCmd(this.nodeSetupConfig, this.configPath, this.dataPath, false);
+  }
+
+  private getZombieRole() {
+    const { zombieRole, validator } = this.nodeSetupConfig;
+
+    if (zombieRole) {
+      return zombieRole;
+    } else if (validator) {
+      return "authority";
+    }
+
+    return "full-node";
+  }
+
+  protected getEnv() {
+    const { env } = this.nodeSetupConfig;
+
+    return env.reduce((memo, item: envVars) => {
+      memo[item.name] = item.value;
+      return memo;
+    }, {} as ProcessEnvironment);
+  }
+
+  protected generateNodeSpec(
+    ports: any[],
+    command: string[],
+    zombieRole: string,
+    env: ProcessEnvironment,
+  ): any {
```

Comment:
Can we avoid `any` as a type? This is one of the pains we have at the moment :)

Reply:
Done